test/bandit/http1/request_test.exs

```diff
@@ -483,27 +483,79 @@ defmodule HTTP1RequestTest do
       assert response.status == 400
     end
 
-    test "reads a content-length encoded body properly when more of it arrives than we want to read",
+    test "handles the case where we ask for less than is already in the buffer", context do
+      client = SimpleHTTP1Client.tcp_client(context)
+
+      :gen_tcp.send(
+        client,
+        "POST /in_buffer_read HTTP/1.1\r\nhost: localhost\r\ncontent-length: 5\r\n\r\nABCDE"
+      )
+
+      assert {:ok, "200 OK", _, "A,BCDE"} = SimpleHTTP1Client.recv_reply(client)
+    end
+
+    def in_buffer_read(conn) do
+      {:more, first, conn} = Plug.Conn.read_body(conn, length: 1)
+      {:ok, second, conn} = Plug.Conn.read_body(conn)
+      send_resp(conn, 200, "#{first},#{second}")
+    end
+
+    test "handles the case where we ask for more than is already in the buffer", context do
+      client = SimpleHTTP1Client.tcp_client(context)
+
+      :gen_tcp.send(
+        client,
+        "POST /beyond_buffer_read HTTP/1.1\r\nhost: localhost\r\ncontent-length: 5\r\n\r\nAB"
+      )
+
+      Process.sleep(100)
```

Comment:
The intent here is to build up a pretty synthetic situation to force the read logic down specific paths (during the workup of this ticket I'd added explicit `IO.puts` calls to ensure such paths were taken, but we don't have a way to programmatically validate that in tests). Part of that work is ensuring that the server has read a known portion of the request body at each step.

Reply:
Clear, thanks!

jfyi I think it's possible to trace with Erlang which functions a process calls. Traces can be simply `assert_received` in ExUnit test. It would be fun to try it out with Elixir once.

Reply:
That *would* be cool! I'm trying to keep tests as covering combinations of externally observable behaviour and not a run through of the internal implementation (ie: the tests are 'what happens if the plug asks for small amounts of data that we already have in buffer', or 'what happens if the plug asks for data that is beyond the scope of the buffer'?) so mandating specific lines of execution isn't *exactly* what we want here.